### DISCUSSION TITLE
Converting the Minimal's VGA to NTSC composite

### DISCUSSION THREAD

**Discussion by nateo87** [slu4coder/Minimal-64x4-Home-Computer#31] [Show and tell]
Just wanted to show off real quick what I've been working on. For some reason, the earlier topic about having an automatic boot-reset got me thinking about using this little wonder with a CRT. Here in the states, a couple computers were able to come up with color graphics by cheating the wonky way my country used to do color broadcasting. Black and white pixels were put very close together (but not touching) and because these were not precise dots by the time they were squeezed into the composite signal, the television would misinterpret it as a color. Since the Minimal is 1-bit black and white, I had to know if it was capable of the same sort of magic.

So here's what I got so far - the picture is stable. Using 74HC14 schmitt trigger delays using logic to select the difference, I was able to get both horizontal and vertical sync within tolerance of the NTSC standard. I also XORed the two sync signals and ANDed them with the main video signal so that the computer would never generate video during those crucial spots.

The main trick I'm working through now is cutting down the video resolution by fitting it into a window consisting of divisions of the horizontal and vertical frequency. I've got it so that the image isn't doubling up horizontally, but I'm still figuring out the vertical. I'll also probably have to throw in a PWM for the video so it fills out more of the screen. 
![20240928_005620](https://github.com/user-attachments/assets/155431f5-d201-4b59-a6d1-f3262a2c3c89)

**Comment by nateo87** [comment]
Got pretty far with it, managed even to get the right window, but the image still scrolls slowly and blinks. The pixel clock just moves way too fast for NTSC. This could probably be tackled by making a few changes on the board (a cut trace here, a bodge there, sticking in an NTSC-friendly frequency in the 74HC166), but that'll have to wait for another day. For now, I'm just pleased I got that far at all without taking a razor blade to the board. 

Any suggestions from fellow tinkerers on simple-ish solution might be?

**Reply by slu4coder** [reply]
Hey Nate, great project :-) At some stage I thought about adding a composite output but it wasn't 'minimal' enough, so I kept it simple with VGA. I fear the Minimal 64x4 is in a way not suited very well to using it outside what its designed to do since the circuit can *barely* do what it does. Timing is super critical. So what you have accomplished so far is way more than I thought it was possible :-)

**Reply by nateo87** [reply]
You say "it can barely do what it does", I say: it's very elegantly designed! I'm having so much fun with this computer. Might even make a lil game for it as my first foray into assembler - I got a little arrow figure to wrap around the screen!

Something I discovered very quickly in this side quest of mine is just how critical all that timing is. VGA is also _so much faster_ than NTSC. By the time NTSC is finishing its first horizontal scan line, 60hz VGA is already starting its third. I'm still fairly certain I can pull this off, but it's going to have to require a full frame buffer, which is something I was originally hoping to avoid. I thought I might be able to do a sort of FIFO memory using alternating D-latch registers, but even if the VRAM was sampled at 50%, composite sampling would still be moving too slow. Luckily, there's a lot of time available in the composite signal during the vertical blank, and would be plenty of time to take a fairly wide sample of the onboard VRAM, in addition to the extra time gained by the windowing of the visible screen real estate.

Anyways, that'll be for another day - other projects call to me!

**Comment by nateo87** [comment]
Picking this up again - I've done a rough sketch in KiCad and I believe I've found a way to make it work. And yes, it's recursive, unecessary, and probobaly over-engineered, but I just can't help myself. This adapter board would be used in conjunction with the TTL B&W composite video generator I designed.
![schm_prelim](https://github.com/user-attachments/assets/a64ab482-2877-4b01-87e1-024c03f8e06f)

**Reply by nateo87** [reply]
Oh, and the reason there's two 62256s to the left is that one of them is for a flat cable with DIP-28 plugs to connect the adapter board to the mainboard, and the other is a socket for the original, main VGA VRAM. I included a 2x14 pin head socket for.... some reason? I dunno, I'm still messing around :-P
